### PR TITLE
Соловьев А.Ю. 381806-1

### PR DESCRIPTION
**Pull request by Alexandr-Solovev**

### CODE REVIEW

File: main.cpp [hunk 1]
```diff
@@ -19,13 +19,34 @@ using namespace clang::tooling;
...
+	auto re_range = CharSourceRange::getCharRange (result_item->getLParenLoc(), result_item->getSubExprAsWritten()->getBeginLoc());
+
+	StringRef type_string = 
+              Lexer::getSourceText(CharSourceRange::getTokenRange(result_item->getLParenLoc().getLocWithOffset(1),
+                                   result_item->getRParenLoc().getLocWithOffset(-1)),
+		                           source_manager, Result.Context->getLangOpts());
+
+	std::string str = ("static_cast<" + type_string + ">").str();
+
+	const Expr* sub_expr = result_item->getSubExprAsWritten()->IgnoreImpCasts();
```
Comment (eshulankina) [inline main.cpp:38]: Please describe what happens if change the code as follows:  
```suggestion
	const Expr* sub_expr = result_item->getSubExpr();
```
What is the difference between `getSubExprAsWritten()` and `getSubExpr()`? Please provide an example.  
What is the purpose of using the `IgnoreImpCasts()` method?

Comment (Alexandr-Solovev) [reply main.cpp:38]: 1)`const Expr* sub_expr = result_item->getSubExprAsWritten()->IgnoreImpCasts();`
Output: `int main() {
    float f;
    int i = static_cast<int>(f);
    return 0;
}`
`const Expr* sub_expr = result_item->getSubExpr();`
Output: `int main() {
    float f;
    int i = static_cast<int>(f);
    return 0;
}`
Ничего не изменилось c:
3) `IgnoreImpCasts()` пропускает неявное приведение типов, здесь использовать не обязательно.
2) `getSubExprAsWritten()` Retrieve the cast subexpression as it was written in the source code, looking through any implicit casts or other intermediate nodes introduced by semantic analysis.
`getSubExpr()` take expr after semantic analysis  `->` result with `getSubExpr()` may be incorrect in the case of semantic analysis)

Comment (eshulankina) [reply main.cpp:38]: If you try to print the AST of the `test.cpp` example, you will see that you have several `ImplicitCastExpr` nodes:
image

Depending on which approaches you use to get the subexpression node, you will get different result nodes for the `test.cpp` example:

Code:
```cpp
// getSubExpr() retrieves the first cast subexpression
llvm::outs() << "getSubExpr(): " << castNode->getSubExpr()->getStmtClassName();

// getSubExprAsWritten() retrieves the cast subexpression as it was written in the source
// code, looking through any implicit casts or other intermediate nodes
// introduced by semantic analysis.
llvm::outs() << "getSubExprAsWritten(): " << castNode->getSubExprAsWritten()->getStmtClassName();

// getSubExpr() retrieves the first cast subexpression.
// ignoreImpCasts() skips past any implicit casts which might surround this expression until
// reaching a fixed point. It skips the following AST nodes: ImplicitCastExpr, FullExpr.
llvm::outs() << "getSubExpr()->ignoreImpCasts(): " << castNode->getSubExpr()->IgnoreImpCasts()->getStmtClassName();
```
Output:
```bash
 getSubExpr(): ImplicitCastExpr 
 getSubExprAsWritten(): DeclRefExpr  
 getSubExpr()->ignoreImpCasts(): DeclRefExpr
```

File: main.cpp [hunk 2]
```diff
@@ -19,13 +19,34 @@ using namespace clang::tooling;
...
 
     void run(const MatchFinder::MatchResult &Result) override {
-        // Your code goes here
+    const auto* result_item = Result.Nodes.getNodeAs<CStyleCastExpr>("cast");
+	SourceManager& source_manager = *Result.SourceManager;
+
+	auto re_range = CharSourceRange::getCharRange (result_item->getLParenLoc(), result_item->getSubExprAsWritten()->getBeginLoc());
+
+	StringRef type_string = 
+              Lexer::getSourceText(CharSourceRange::getTokenRange(result_item->getLParenLoc().getLocWithOffset(1),
```
Comment (eshulankina) [inline main.cpp:32]: Please describe the difference between `getTokenRange()` and `getCharRange()` with examples.

Comment (Alexandr-Solovev) [reply main.cpp:32]: `getTokenRange()` возвращает диапазон до первого символа последнего токена
` getCharRange() `возвращает диапазон до последнего символа, включая его

Добавил в main.cpp
```
auto charRange = Lexer::getAsCharRange(
    CharSourceRange::getCharRange(result_item->getLParenLoc().getLocWithOffset(1),
    result_item->getRParenLoc().getLocWithOffset(-2)),
    source_manager, Result.Context->getLangOpts());

auto tokenRange = Lexer::getAsCharRange(
    CharSourceRange::getTokenRange(result_item->getLParenLoc().getLocWithOffset(1),
    result_item->getRParenLoc().getLocWithOffset(-2)),
    source_manager, Result.Context->getLangOpts());
    llvm::outs() << "\n char range start: " << charRange.getBegin().printToString(source_manager) << ",\n char range end: " << charRange.getEnd().printToString(source_manager) << "\n";
    llvm::outs() << "\n token range start: " << tokenRange.getBegin().printToString(source_manager) << ",\n token range end: " << tokenRange.getEnd().printToString(source_manager) << "\n";
```

Получил
```
 char range start: /mnt/d/Programming/UNN-C-Style-Checker/test/test.cpp:3:14,
 char range end: /mnt/d/Programming/UNN-C-Style-Checker/test/test.cpp:3:15

 token range start: /mnt/d/Programming/UNN-C-Style-Checker/test/test.cpp:3:14,
 token range end: /mnt/d/Programming/UNN-C-Style-Checker/test/test.cpp:3:17
int main() {
    float f;
    int i = static_cast<int>();
    return 0;
}
```

**Review by eshulankina — APPROVED** [review]

**Pull request closed without merging**